quickwit/quickwit-control-plane/src/scheduler.rs

```diff
@@ -17,37 +17,585 @@
+///   scheduling on each event. TODO(fmassot): this can be refined by adding some relevant info to
+///   the event, example: the creation of a source of type `void` should not trigger a sheduling.
+/// - [`RefreshPlanLoop`]: this event is scheduled every [`REFRESH_PLAN_LOOP_INTERVAL`] and triggers
+///   a scheduling. Due to network issues, a control plane will not always receive the gRPC events
+///   [`NotifyIndexChangeRequest`] and thus will not be aware of index changes in the metastore.
+///   TODO(fmassot): to avoid a scheduling on each [`RefreshPlanLoop`], we can store in the
+///   scheduler state a metastore version number that will be compared to the number stored in the
+///   metastore itself.
+/// - [`ControlPlanLoop`]: this event is sheduled every [`HEARTBEAT`] and control if the `desired
+///   plan`, that is the last applied [`PhysicalIndexingPlan`] by the scheduler, and the `running
+///   plan`, that is the indexing tasks running on all indexers and retrieved from the chitchat
+///   state, are the same. If not, the scheduler will trigger a scheduling.
+///
+/// Finally, in order to give the time for each indexer to run their indexing tasks, the control
+/// phase will wait at least [`MIN_DURATION_BETWEEN_SCHEDULING_SECS`] before comparing the desired
+/// plan with the running plan.
 pub struct IndexingScheduler {
     cluster: Arc<Cluster>,
     metastore: Arc<dyn Metastore>,
+    indexing_client_pool: IndexingClientPool,
+    state: IndexingSchedulerState,
 }
 
 #[async_trait]
 impl Actor for IndexingScheduler {
-    type ObservableState = ();
+    type ObservableState = IndexingSchedulerState;
 
-    fn observable_state(&self) -> Self::ObservableState {}
+    fn observable_state(&self) -> Self::ObservableState {
+        self.state.clone()
+    }
 
     fn name(&self) -> String {
-        "IndexerScheduler".to_string()
+        "IndexingScheduler".to_string()
     }
 
-    async fn initialize(&mut self, _: &ActorContext<Self>) -> Result<(), ActorExitStatus> {
+    async fn initialize(&mut self, ctx: &ActorContext<Self>) -> Result<(), ActorExitStatus> {
+        self.handle(RefreshPlanLoop, ctx).await?;
+        ctx.schedule_self_msg(HEARTBEAT, ControlPlanLoop).await;
         Ok(())
     }
 }
 
 impl IndexingScheduler {
-    pub fn new(cluster: Arc<Cluster>, metastore: Arc<dyn Metastore>) -> Self {
-        Self { cluster, metastore }
+    pub fn new(
+        cluster: Arc<Cluster>,
+        metastore: Arc<dyn Metastore>,
+        indexing_client_pool: IndexingClientPool,
+    ) -> Self {
+        Self {
+            cluster,
+            metastore,
+            indexing_client_pool,
+            state: IndexingSchedulerState::default(),
+        }
+    }
+
+    async fn schedule_indexing_plan(&mut self) -> anyhow::Result<()> {
+        let indexers = self.get_indexers_from_cluster_state().await;
+        if indexers.is_empty() {
```

Comment:
so that's where the contract was enforced. That's two function calls away. It is a bit scary.